ctxtool/cancel.go

```diff
@@ -0,0 +1,37 @@
+package ctxtool
+
+import "context"
+
+// AutoCancel collects cancel functions to be executed at the end of the
+// function scope.
+//
+// Example:
+//   var ac AutoCancel
+//   defer ac.Cancel()
+//   ctx := ac.With(context.WithCancel(context.Background()))
+//   ctx := ac.With(context.WithTimeout(ctx, 5 * time.Second))
+//   ... // do something with ctx
+type AutoCancel struct {
+	funcs []context.CancelFunc
+}
+
+// Cancel calls all registered cancel functions in reverse order.
+func (ac *AutoCancel) Cancel() {
+	for _, fn := range ac.funcs {
+		defer fn()
+	}
+}
+
+// Add adds a new cancel function to the AutoCancel. The function will be run
+// before any other already registered cancel function.
+func (ac *AutoCancel) Add(fn context.CancelFunc) {
+	ac.funcs = append(ac.funcs, fn)
+}
+
+// With is used to wrap a Context constructer call that returns a context and a
+// cancel function.  The cancel function is automatically added to AutoCancel
+// and the original context is returned as is.
+func (ac *AutoCancel) With(ctx context.Context, cancel context.CancelFunc) context.Context {
+	ac.Add(cancel)
+	return ctx
+}
```

Comment:
Oh this is interesting in what context you want to use this?

Reply:
One of the problem with contexts is, that you need to clean them up in case you leave the scope without the parent context being cancelled. Context that return a cancel function require you to call cancel to cleanup. This gives you code like this:

```
ctx, cancel := context.WithX(...)
defer cancel()

ctx, cancel = context.WithX(ctx, ...)
defer cancel()
```

or more annoying:

```
ctx := context.WithX()

var cancel context.CancelFunc
ctx, cancel = context.WithY(...)
defer cancel()
```

The `With` method allows me to capture the cancel function and call 'Cancel' only once. Case 1 becomes:

```
var ac AutoCancel
defer ac.Cancel()

ctx := ac.With(context.WithX(...))
ctx = ac.With(context.WithY(ctx, ...))
```

Case 2 becomes:

```
var ac AutoCancel
defer ac.Cancel()

ctx := context.WithX()
ctx = ac.With(context.WithY(ctx, ...))
```

This works because go allows me to wrap a function return directory if the arguments and multi-return values match in number and types. In this sense the AutoCloser is mostly for convenience.

Normally one should not hold on a context, because a contexts scope is the current function. But in case one only wants to use context for cancellation but integrate it into a subsystem not makind use of context, and wanting to shut-down all workers once a stop signal is received:

```
type controller struct {
  ac AutoCancel
  ...
}

func (c controller) Stop() {
  c.ac.Cancel()
}

func (c controller) AddContext(fn func(ctx, ...)) {
  ctx := c.ac.With(context.WithCancel(context.Background()))
  go fn(ctx, ...)
}

func (c controller) AddRunner(runner interface{ Start(); Stop() }) {
  c.ac.Add(runner.Stop)
  go runner.Start()
}
```

Reply:
much simpler indeed.